xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/store/XWikiHibernateStore.java

```diff
@@ -2066,59 +2071,60 @@ public String getName()
             @Override
             public List<Event> getEvents()
             {
-                return Collections.<Event>singletonList(this.ev);
+                return EVENT_LIST;
             }
 
             @Override
             public void onEvent(Event event, Object source, Object data)
             {
-                if ("logout".equals(((ActionExecutingEvent) event).getActionName())) {
-                    final XWikiContext ctx = (XWikiContext) data;
-                    if (ctx.getUserReference() != null) {
-                        releaseAllLocksForCurrentUser(ctx);
-                    }
+                UserUnauthenticatedEvent userUnauthenticatedEvent = (UserUnauthenticatedEvent) event;
+                if (userUnauthenticatedEvent.getUserReference() != null) {
+                    DocumentReference userDoc = XWikiHibernateStore.this.userReferenceSerializer.serialize(
+                        userUnauthenticatedEvent.getUserReference());
+                    releaseAllLocksForUser(userDoc, (XWikiContext) data);
```

Comment:
Why does this release all locks? Shouldn't this only release locks of the current session of the user? I mean when I log out on my phone, this shouldn't release locks that I'm holding on my desktop browser, should it?

Reply:
We don't really have the concept of session right now in the locks, but yes ideally they should.

Reply:
That would be an improvment to add in the future I think, the expectation of current mechanism is that it does release all locks

Reply:
Could be done at the same time as moving the locks to solr for better perf as Thomas suggested.

Reply:
Storing a reference to all active locks of the user in the session would solve both the session and the performance problem.

Reply:
> Storing a reference to all active locks of the user in the session would solve both the session and the performance problem.

It would probably be enough for the performance problem, yes (there is very little chance that the user has locks on many wikis at the same time in practice).

Reply:
I'd do that as part of another improvment ticket: IMO it's out of the scope of current PR.

Reply:
It's kind of the same scope as the `for` you added, the session references was another way to fix this problem.